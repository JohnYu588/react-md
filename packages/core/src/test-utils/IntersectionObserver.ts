export class IntersectionObserverMock implements IntersectionObserver {
  root: Document | Element | null;
  rootMargin: string;
  thresholds: readonly number[];

  elements: Set<Element>;

  constructor(
    public callback: IntersectionObserverCallback,
    options: IntersectionObserverInit = {}
  ) {
    this.root = options.root || null;
    this.rootMargin = options.rootMargin || "";
    this.thresholds =
      typeof options.threshold === "number"
        ? [options.threshold]
        : options.threshold ?? [];

    this.elements = new Set();
  }

  observe(target: Element): void {
    this.elements.add(target);

    this.callback(this.takeRecords(), this);
  }

  unobserve(target: Element): void {
    this.elements.delete(target);
  }

  takeRecords(): IntersectionObserverEntry[] {
    return [...this.elements].map<IntersectionObserverEntry>((target) => ({
      time: Date.now(),
      target,
      boundingClientRect: target.getBoundingClientRect(),
      intersectionRatio: 0,
      intersectionRect: target.getBoundingClientRect(),
      isIntersecting: false,
      rootBounds:
        this.root && "getBoundingClientRect" in this.root
          ? this.root.getBoundingClientRect()
          : null,
    }));
  }

  disconnect(): void {
    this.elements.clear();
  }
}
