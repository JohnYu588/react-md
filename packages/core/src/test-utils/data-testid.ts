/// <reference types="react" />

declare module "react" {
  interface HTMLAttributes<T>
    extends React.AriaAttributes,
      React.DOMAttributes<T> {
    "data-testid"?: string | number;
  }
}
