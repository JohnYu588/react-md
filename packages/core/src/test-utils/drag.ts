import { fireEvent } from "@testing-library/dom";
import type { MouseEvent } from "react";
import { wait } from "../utils/wait.js";

interface XYCoords {
  x: number;
  y: number;
}

interface BaseDragOptions {
  to?: XYCoords | Element;
  delta?: XYCoords;
  steps?: number;
  duration?: number;
}

type DragOptions = BaseDragOptions &
  ({ to: XYCoords | Element; delta?: never } | { delta: XYCoords; to?: never });

const getElementClientCenter = (element: Element): XYCoords => {
  const { left, top, width, height } = element.getBoundingClientRect();
  return {
    x: left + width / 2,
    y: top + height / 2,
  };
};

const getCoords = (elementOrCoords: Element | XYCoords): XYCoords =>
  "x" in elementOrCoords && "y" in elementOrCoords
    ? elementOrCoords
    : getElementClientCenter(elementOrCoords);

/**
 * This is internal for now. No idea if it's actually worthwhile though since
 * you have to mock the `getBoundingClientRect()` to work.
 *
 * @internal
 * @see https://github.com/testing-library/testing-library-docs/blob/c20f8a8931e9f0a158fda9f01bbea8cb45ed1651/docs/example-drag.mdx
 */
export async function drag(
  element: Element,
  options: DragOptions
): Promise<void> {
  const { delta, to: inTo, steps = 20, duration = 500 } = options;
  const from = getElementClientCenter(element);
  const to = delta
    ? {
        x: from.x + delta.x,
        y: from.y + delta.y,
      }
    : getCoords(inTo);

  const step: XYCoords = {
    x: (to.x - from.x) / steps,
    y: (to.y - from.y) / steps,
  };

  const currentEvent = {
    clientX: from.x,
    clientY: from.y,
    button: 0,
    altKey: false,
    ctrlKey: false,
    metaKey: false,
    shiftKey: false,
  } satisfies Partial<MouseEvent>;

  fireEvent.mouseEnter(element, currentEvent);
  fireEvent.mouseOver(element, currentEvent);
  fireEvent.mouseMove(element, currentEvent);
  fireEvent.mouseDown(element, currentEvent);
  fireEvent.dragStart(element, currentEvent);

  const delay = duration / steps;
  for (let i = 0; i < steps; i++) {
    currentEvent.clientX += step.x;
    currentEvent.clientY += step.y;
    await wait(delay);
    fireEvent.mouseMove(element, currentEvent);
  }

  fireEvent.mouseUp(element, currentEvent);
  fireEvent.dragEnd(element, currentEvent);
}
