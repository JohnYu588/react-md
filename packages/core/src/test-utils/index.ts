export * from "@testing-library/react";
export * from "@testing-library/user-event";
export * from "./mocks/match-media.js";
export * from "./render.js";
export * from "./utils/resize-observer.js";
