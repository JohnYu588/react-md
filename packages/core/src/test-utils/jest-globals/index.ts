export * from "./match-media.js";
export * from "./resize-observer.js";
export * from "./timers.js";
