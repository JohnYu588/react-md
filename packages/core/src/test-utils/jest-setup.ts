import { beforeEach } from "@jest/globals";
import { INTERACTION_CONFIG } from "../interaction/config.js";

beforeEach(() => {
  // set the mode to `none` in tests since ripples require
  // `getBoundingClientRect()` to create correct CSS. You'll either see warnings
  // in the console around invalid css values or `NaN`.
  INTERACTION_CONFIG.mode = "none";
});
