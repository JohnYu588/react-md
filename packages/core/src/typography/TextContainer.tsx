import { cnb } from "cnbuilder";
import type { HTMLAttributes } from "react";
import { forwardRef } from "react";
import { bem } from "../utils";

const styles = bem("rmd-text-container");

/**
 * A union of the available text container sizes. One of these values must be
 * chosen to help set the max width for text.
 */
export type TextContainerSize = "auto" | "mobile" | "desktop";

export interface TextContainerClassNameOptions {
  /**
   * @defaultValue `"auto"`
   * @see {@link TextContainerSize}
   */
  size?: TextContainerSize;

  /**
   * An optional className to merge with typography text container styles.
   */
  className?: string;
}

/**
 * @example
 * Simple Example
 * ```tsx
 * import { getTextContainerClassName, Typography } from "@react-md/core";
 *
 * function Example() {
 *   return (
 *     <main className={getTextContainerClassName()}>
 *       <Typography type="headline-1">Heading</Typography>
 *       <Typography>
 *         Pretend this is a giant paragraph of text that wraps multiple lines.
 *       </Typography>
 *       <Typography>
 *         Pretend this is another giant paragraph of text that wraps multiple
 *         lines.
 *       </Typography>
 *     </main>
 *   ):
 * }
 * ```
 *
 * @param options - An optional object of options used to create the text
 * container class name.
 * @returns the text container class name
 * @see {@link TextContainer}
 * @remarks \@since 6.0.0
 */
export function textContainer(
  options: TextContainerClassNameOptions = {}
): string {
  const { size = "auto", className } = options;

  return cnb(styles({ [size]: true }), className);
}

export interface TextContainerProps
  extends HTMLAttributes<HTMLDivElement>,
    TextContainerClassNameOptions {}

/**
 * This component should be used to render text based content with an
 * appropriate max line length to optimize legibility.
 *
 * @example
 * Simple Example
 * ```tsx
 * import { TextContainer, Typography } from "@react-md/core";
 *
 * function Example() {
 *   return (
 *     <TextContainer>
 *       <Typography type="headline-1">Heading</Typography>
 *       <Typography>
 *         Pretend this is a giant paragraph of text that wraps multiple lines.
 *       </Typography>
 *       <Typography>
 *         Pretend this is another giant paragraph of text that wraps multiple
 *         lines.
 *       </Typography>
 *     </TextContainer>
 *   );
 * }
 * ```
 */
export const TextContainer = forwardRef<HTMLDivElement, TextContainerProps>(
  function TextContainer(props, ref) {
    const { size = "auto", className, children, ...remaining } = props;

    return (
      <div
        {...remaining}
        ref={ref}
        className={textContainer({ size, className })}
      >
        {children}
      </div>
    );
  }
);
