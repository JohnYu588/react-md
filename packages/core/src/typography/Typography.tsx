import {
  forwardRef,
  type ElementType,
  type HTMLAttributes,
  type ReactElement,
} from "react";
import {
  typography,
  type TypographyClassNameOptions,
  type TypographyType,
} from "./typographyStyles.js";

/**
 * A union of the default supported elements that the `Typography` component can
 * be rendered as. This is mostly used for adding the correct `HTMLAttributes`
 * and enabling the forward ref.
 *
 * @since 4.0.0
 */
export type TypographyHTMLElement =
  | HTMLHeadingElement
  | HTMLParagraphElement
  | HTMLSpanElement
  | HTMLDivElement
  | HTMLAnchorElement
  | HTMLBodyElement
  | HTMLHtmlElement;

/** @since 6.0.0 */
export type CustomTypographyComponent = ElementType<
  HTMLAttributes<TypographyHTMLElement> & { className: string }
>;

/** @internal */
function getComponent(
  as: CustomTypographyComponent | undefined,
  type: TypographyType
): ElementType {
  if (as) {
    return as;
  }

  switch (type) {
    case "headline-1":
      return "h1";
    case "headline-2":
      return "h2";
    case "headline-3":
      return "h3";
    case "headline-4":
      return "h4";
    case "headline-5":
      return "h5";
    case "headline-6":
    case "subtitle-1":
    case "subtitle-2":
      return "h6";
    case "body-1":
    case "body-2":
      return "p";
    case "caption":
      return "caption";
    default:
      return "span";
  }
}

export interface TypographyProps
  extends HTMLAttributes<TypographyHTMLElement>,
    TypographyClassNameOptions {
  /**
   * The component to render as when the children are not a render function. If
   * this prop is omitted, the component will be determined by the `type` prop
   * where:
   *
   * - `"headline-1" -> <h1>`
   * - `"headline-2" -> <h2>`
   * - `"headline-3" -> <h3>`
   * - `"headline-4" -> <h4>`
   * - `"headline-5" -> <h5>`
   * - `"headline-6" -> <h6>`
   * - `"subtitle-1" -> <h5>`
   * - `"subtitle-2" -> <h6>`
   * - `"body-1"     -> <p>`
   * - `"body-2"     -> <p>`
   * - `"caption"    -> <caption>`
   * - `"overline"   -> <span>`
   */
  as?: CustomTypographyComponent;
}

/**
 * **Server Component**
 *
 * Render text with one of the material design typography styles applied and
 * optional styles like font-weight, font-style, text color, etc.
 *
 * @example All Example
 * ```tsx
 * import { Typography } from "@react-md/core":
 *
 * export function Example() {
 *   return (
 *    <>
 *      <Typography type="headline-1">Headline 1</Typography>
 *      <Typography type="headline-2">Headline 2</Typography>
 *      <Typography type="headline-3">Headline 3</Typography>
 *      <Typography type="headline-4">Headline 4</Typography>
 *      <Typography type="headline-5">Headline 5</Typography>
 *      <Typography type="headline-6">Headline 6</Typography>
 *      <Typography type="subtitle-1">Subtitle 1</Typography>
 *      <Typography type="subtitle-2">Subtitle 2</Typography>
 *      <Typography>
 *        A paragraph of text.
 *      </Typography>
 *      <Typography type="body-1">
 *        A paragraph of text.
 *      </Typography>
 *      <Typography type="body-2">
 *        Another paragraph of text.
 *      </Typography>
 *      <Typography type="caption" component="h5">
 *        Caption text
 *      </Typography>
 *      <Typography type="overline" component="h5">
 *        Overline text
 *      </Typography>
 *    </>
 *   ):
 * }
 * ```
 */
export const Typography = forwardRef<TypographyHTMLElement, TypographyProps>(
  function Typography(props, ref): ReactElement {
    const {
      as,
      type = "body-1",
      className,
      margin,
      fontStyle,
      fontWeight,
      textAlign,
      textColor,
      textDecoration,
      textTransform,
      textOverflow,
      children,
      ...remaining
    } = props;

    const Component = getComponent(as, type);
    return (
      <Component
        {...remaining}
        ref={ref}
        className={typography({
          type,
          margin,
          fontStyle,
          fontWeight,
          textAlign,
          textColor,
          textDecoration,
          textTransform,
          textOverflow,
          className,
        })}
      >
        {children}
      </Component>
    );
  }
);
