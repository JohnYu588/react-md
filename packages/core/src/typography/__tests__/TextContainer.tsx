import { createRef } from "react";
import { render } from "../../test-utils";

import { textContainer, TextContainer } from "../TextContainer";
import { Typography } from "../Typography";

describe("TextContainer", () => {
  it("should render correctly and apply the correct class names", () => {
    const { getByTestId, rerender } = render(
      <TextContainer data-testid="container">
        <Typography>Content</Typography>
      </TextContainer>
    );

    const container = getByTestId("container");
    expect(container).toHaveClass("rmd-text-container");
    expect(container).toHaveClass("rmd-text-container--auto");
    expect(container).not.toHaveClass("rmd-text-container--mobile");
    expect(container).not.toHaveClass("rmd-text-container--desktop");
    expect(container).toMatchSnapshot();

    rerender(
      <TextContainer data-testid="container" size="mobile">
        <Typography>Content</Typography>
      </TextContainer>
    );
    expect(container).toHaveClass("rmd-text-container");
    expect(container).not.toHaveClass("rmd-text-container--auto");
    expect(container).toHaveClass("rmd-text-container--mobile");
    expect(container).not.toHaveClass("rmd-text-container--desktop");
    expect(container).toMatchSnapshot();

    rerender(
      <TextContainer data-testid="container" size="desktop">
        <Typography>Content</Typography>
      </TextContainer>
    );
    expect(container).toHaveClass("rmd-text-container");
    expect(container).not.toHaveClass("rmd-text-container--auto");
    expect(container).toHaveClass("rmd-text-container--desktop");
    expect(container).not.toHaveClass("rmd-text-container--mobile");
    expect(container).toMatchSnapshot();

    rerender(
      <TextContainer data-testid="container" className="custom-1 custom-2">
        <Typography>Content</Typography>
      </TextContainer>
    );

    expect(container).toHaveClass("rmd-text-container");
    expect(container).toHaveClass("rmd-text-container--auto");
    expect(container).not.toHaveClass("rmd-text-container--mobile");
    expect(container).not.toHaveClass("rmd-text-container--desktop");
    expect(container).toHaveClass("custom-1");
    expect(container).toHaveClass("custom-2");
    expect(container).toMatchSnapshot();
  });

  it("should forward the ref to the div element", () => {
    const ref = createRef<HTMLDivElement>();
    render(<TextContainer ref={ref} />);

    expect(ref.current).toBeInstanceOf(HTMLDivElement);
  });

  describe("styling utility class", () => {
    it("should be callable without any arguments", () => {
      expect(textContainer()).toMatchSnapshot();
    });
  });
});
