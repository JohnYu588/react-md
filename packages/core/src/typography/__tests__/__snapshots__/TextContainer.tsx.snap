// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextContainer should render correctly and apply the correct class names 1`] = `
<div
  class="rmd-text-container rmd-text-container--auto"
  data-testid="container"
>
  <p
    class="rmd-typography rmd-typography--body-1"
  >
    Content
  </p>
</div>
`;

exports[`TextContainer should render correctly and apply the correct class names 2`] = `
<div
  class="rmd-text-container rmd-text-container--mobile"
  data-testid="container"
>
  <p
    class="rmd-typography rmd-typography--body-1"
  >
    Content
  </p>
</div>
`;

exports[`TextContainer should render correctly and apply the correct class names 3`] = `
<div
  class="rmd-text-container rmd-text-container--desktop"
  data-testid="container"
>
  <p
    class="rmd-typography rmd-typography--body-1"
  >
    Content
  </p>
</div>
`;

exports[`TextContainer should render correctly and apply the correct class names 4`] = `
<div
  class="rmd-text-container rmd-text-container--auto custom-1 custom-2"
  data-testid="container"
>
  <p
    class="rmd-typography rmd-typography--body-1"
  >
    Content
  </p>
</div>
`;

exports[`TextContainer styling utility class should be callable without any arguments 1`] = `"rmd-text-container rmd-text-container--auto"`;
